flink-runtime/src/main/java/org/apache/flink/runtime/state/heap/HeapKeyedStateBackend.java

```diff
@@ -412,9 +421,15 @@ private void restorePartitionedState(Collection<KeyGroupsStateHandle> state) thr
 					}
 				}
 
-				for (Tuple2<Integer, Long> groupOffset : keyGroupsHandle.getGroupRangeOffsets()) {
+				for (Tuple2<Integer, Long> groupOffset : keyGroupsStateHandle.getGroupRangeOffsets()) {
 					int keyGroupIndex = groupOffset.f0;
 					long offset = groupOffset.f1;
+
+					// Skip those key groups that don't belong to the backend.
+					if (!keyGroupRange.contains(keyGroupIndex)) {
```

Comment:
I think this is no longer required, because we are now "cutting out" the right key-groups from each state handle in the `StateAssignmentOperation`, using the `KeyedStateHandle::getIntersection(...)` method. In fact, I think that receiving a key-group that is not in the backend's range could now be considered as an error. We could rewrite this as a precondition check, or an assert.